xla/client/xla_builder.cc

```diff
@@ -112,6 +112,22 @@ bool InstrIsSetBound(const HloInstructionProto* instr_proto) {
   return false;
 }
 
+bool IsDISCBackend() {
+  char* env = getenv("DISC_DEVICE");
+  return env != nullptr;
+}
+
+StatusOr<XlaOp> CreateShapeTensor(const std::vector<XlaOp>& size_ops) {
+  CHECK(!size_ops.empty()) << "size_ops should not be empty";
+  auto front = size_ops.front();
+  TF_ASSIGN_OR_RETURN(const Shape* shape, front.builder()->GetShapePtr(front));
+  Shape result_shape =
+      ShapeUtil::MakeShape(shape->element_type(), {size_ops.size()});
+  return CustomCall(front.builder(), "tensor.from_elements",
```

Comment:
Here, we cannot create tensor dialect ops; we can only create HLO ops. It’s actually better not to use `tensor.from_elements`, but to create shape tensors using `reshape` + `concat`. This is because the op that gets the dimension size returns a scalar tensor, which, when used as a parameter for `tensor.from_elements`, will generate unintended tensors.